rote/minitermite/doc/term2src.1
.\" Man page generated from reStructeredText.
.
.TH MINI-TERMITE 1 "2012-07-10" "1.1" "Compilers and Programming Languages"
.SH NAME
mini-Termite \- Term-based intermediate representation for source code
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" -*- rst -*-
.
.SH SYNOPSIS
.sp
term2src [OPTION]... [FILE.term]
.SH DESCRIPTION
.sp
Unparse a term file to its original source representation.
.INDENT 0.0
.TP
.B Options:
.INDENT 7.0
.TP
.B \-o, \-\-output sourcefile.c
.
If specified, the contents of all files will be concatenated
into the sourcefile.
.TP
.B \-s, \-\-suffix \(aq.suffix\(aq  Default: \(aq.unparsed\(aq
.
Use the original file names with the additional suffix.
.UNINDENT
.INDENT 7.0
.TP
.BI \-d,  \-\-dir \ DIRECTORY
.
Create the unparsed files in DIRECTORY.
.UNINDENT
.INDENT 7.0
.TP
.B \-\-dot
.
Create a dotty graph of the syntax tree.
.TP
.B \-\-pdf
.
Create a PDF printout of the syntax tree.
.TP
.B \-\-stratego
.
Create term output compatible with the Stratego/XT.
.UNINDENT
.UNINDENT
.sp
This program was built against minitermite 1.1,
please report bugs to <\fI\%adrian@complang.tuwien.ac.at\fP>.
.SH About
.sp
Minitermite is a light\-weight library that transforms C, C++, and
Fortran programs into a term\-based representation of their abstract
syntax trees (AST) and vice versa.
.SH Related work, Provenance
.sp
Minitermite is a spin\-off (friendly fork) of Termite [1] which is
distributed as part of the SATIrE program analysis framework [2] from
the Computer Languages group at Vienna University of Technology,
Austria.
.sp
[1] \fI\%http://www.complang.tuwien.ac.at/adrian/termite\fP
[2] \fI\%http://www.complang.tuwien.ac.at/satire\fP
.SH Authors
.sp
The corresponding author and maintainer is Adrian Prantl. The original
Termite was based on an earlier prototype by Christoph Bonitz and
Markus Schordan.  Gergo Barany contributed the term_lint grammar
checker, as well as support for the ICFG and other PAG\-based analyses
and numerous bug fixes and improvements. Several testcases were
contributed by Viktor Pavlu. The majority of the Clang front\-end
integration was contributed by Dietmar Ebner with some help from
Florian Brandner.
.SH Installation instructions
.sp
The following commands should do the job on a UNIX\-like system:
.sp
.nf
.ft C
mkdir build && cd build
\&../configure \-\-prefix=/usr/local \-\-with\-rosedir=/opt/rose \e
             \-\-with\-boostdir=/usr \-\-with\-swi\-prolog=no
make \-j<N> install
make installcheck
.ft P
.fi
.SH Dependencies
.sp
ROSE: \fI\%http://www.rosecompiler.org/\fP
flex, yacc, boost
.sp
Optional: swi\-prolog, clang, doxygen
.SH License
.sp
Just like SATIrE, minitermite is distributed under a permissive
three\-clause BSD license. See COPYING for details.
termlint.pl is covered by the GPLv3.
.\" Generated by docutils manpage writer.
.\" 
.


rote/minitermite/src/term2src.C
/* -*- C++ -*-
Copyright 2006 Christoph Bonitz <christoph.bonitz@gmail.com>
          2007-2008 Adrian Prantl <adrian@complang.tuwien.ac.at>
*/
#include "minitermite.h"
#include <iostream>
#include <stdio.h>
#include <rose.h>
#include "TermToRose.h"
#include <getopt.h>

using namespace std;
using namespace term;

void usage(const char* me) 
{
  cout << "Usage: " << me
       << " [OPTION]... [FILE.term]\n"
       << "Unparse a term file to its original source representation.\n\n"

       << "Options:\n"
       << "  -o, --output sourcefile.c\n"
       << "    If specified, the contents of all files will be concatenated\n"
       << "    into the sourcefile.\n\n"

       << "  -s, --suffix '.suffix'  Default: '.unparsed'\n"

       << "    Use the original file names with the additional suffix.\n\n"

       << "  -d, --dir DIRECTORY\n"
       << "    Create the unparsed files in DIRECTORY.\n\n"

       << "  --dot\n"
       << "    Create a dotty graph of the syntax tree.\n\n"

       << "  --pdf\n"
       << "    Create a PDF printout of the syntax tree.\n\n"

       << "  --stratego\n"
       << "    Create term output compatible with the Stratego/XT.\n\n"
#if HAVE_SWI_PROLOG
       << "  --stl-engine\n"
       << "    Do not use SWI-Prolog to generate term output.\n\n"
#endif

       << "This program was built against "<<PACKAGE_STRING<<",\n"
       << "please report bugs to <"<<PACKAGE_BUGREPORT<<">."

       << endl;

}

int main(int argc, char** argv) {
  //  cout << prote->getRepresentation();
  const char* infile = "";
  const char* outfile= "";
  const char* outdir = ".";
  const char* suffix = ".unparsed";
  int dot_flag = 0;
  int pdf_flag = 0;
  int stratego_flag = 0;
#if HAVE_SWI_PROLOG
  int stl_flag = 0;
#else
  int stl_flag = 1;
#endif
  int version_flag = 0;
  int help_flag = 0;

  while (1) {
    static struct option long_options[] = {
      /* These options set a flag. */
      {"dot", no_argument, &dot_flag, 1},
      {"pdf", no_argument, &pdf_flag, 1},
      {"stratego", no_argument, &stratego_flag, 1},
#if HAVE_SWI_PROLOG
      {"stl-engine", no_argument, &stl_flag, 1},
#endif
      {"version", no_argument, &version_flag, 1},
      {"help", no_argument, &help_flag, 1},
      /* These don't */
      {"dir",    required_argument, 0, 'd'},
      {"output", required_argument, 0, 'o'},
      {"suffix", required_argument, 0, 's'},
      {0, 0, 0, 0}
    };
    /* getopt_long stores the option index here. */
    int option_index = 0;
     
    int c = getopt_long(argc, argv, "d:o:s:", long_options, &option_index);
     
    /* Detect the end of the options. */
    if (c == -1)
      break;
     
    switch (c) {
    case 'd': outdir  = optarg; break;
    case 'o': outfile = optarg; break;
    case 's': suffix  = optarg; break;
     
    default: ;
    }
  }
  if (help_flag) {
    usage(argv[0]);
    return 0;
  }
  if (version_flag) {
    cout << argv[0] << " version " << PACKAGE_VERSION << "\n";
    return 0;
  }
  if (optind < argc) {
    infile = argv[optind];
  } else {
    usage(argv[0]);
    return 1;
  }

  init_termite(argc, argv);

  // Choose the way to parse terms based on the options
  TermFactory* termFactory;
  if (stratego_flag)
    termFactory = new StrategoTermFactory();
  else 
    if (stl_flag)
      termFactory = new STLTermFactory();
#if HAVE_SWI_PROLOG
    else termFactory = new SWIPLTermFactory();
#endif

  TermToRose conv(*termFactory);
  SgNode* p = conv.toRose(infile);

  if (help_flag || version_flag) return 0;

  if (dot_flag) {
    //  Create dot and pdf files
    //  DOT generation (numbering:preoder)
    AstDOTGeneration dotgen;
    dotgen.generateInputFiles((SgProject*)p,AstDOTGeneration::PREORDER);
  }
  if (pdf_flag) {
    //  PDF generation
    AstPDFGeneration pdfgen;
    pdfgen.generateInputFiles((SgProject*)p);
  }
  conv.unparse(outfile, outdir, suffix, p);
  return 0;
}
